circle/src/domain.rs

```diff
@@ -0,0 +1,359 @@
+    }
+}
+
+impl<F: ComplexExtendable> PolynomialSpace for CircleDomain<F> {
+    type Val = F;
+
+    fn size(&self) -> usize {
+        1 << self.log_n
+    }
+
+    fn first_point(&self) -> Self::Val {
+        point_to_univariate(self.shift).unwrap()
+    }
+
+    fn next_point<Ext: ExtensionField<Self::Val>>(&self, x: Ext) -> Option<Ext> {
+        // Only in standard position do we have an algebraic expression to access the next point.
+        if self.is_standard() {
+            let gen = F::circle_two_adic_generator(self.log_n);
+            Some(point_to_univariate(gen.rotate(univariate_to_point(x))).unwrap())
+        } else {
+            None
+        }
+    }
+
+    fn create_disjoint_domain(&self, min_size: usize) -> Self {
+        let mut log_n = log2_ceil_usize(min_size);
+        // Right now we simply guarantee the domain is disjoint by returning a
+        // larger standard position coset, which is fine because we always ask for a larger
+        // domain. If we wanted good performance for a disjoint domain of the same size,
+        // we could change the shift. Also we could support nonstandard twin cosets.
+        assert!(
+            self.is_standard(),
+            "create_disjoint_domain not currently supported for nonstandard twin cosets"
+        );
+        while log_n == self.log_n {
+            log_n += 1;
+        }
+        Self::standard(log_n)
+    }
+
+    fn zp_at_point<Ext: ExtensionField<Self::Val>>(&self, point: Ext) -> Ext {
+        v_n(univariate_to_point(point).real(), self.log_n) - v_n(self.shift.real(), self.log_n)
+    }
+
+    fn selectors_at_point<Ext: ExtensionField<Self::Val>>(
+        &self,
+        point: Ext,
+    ) -> LagrangeSelectors<Ext> {
+        let zeroifier = self.zp_at_point(point);
+        let p = univariate_to_point(point);
+        LagrangeSelectors {
+            is_first_row: zeroifier / v_0(self.shift.conjugate().rotate(p)),
+            is_last_row: zeroifier / v_0(self.shift.rotate(p)),
+            // is_transition: v_0(self.shift.rotate(p)),
+            is_transition: self.shift.rotate(p).real() - Ext::one(),
+            inv_zeroifier: zeroifier.inverse(),
+        }
+    }
+
+    // wow, really slow!
+    #[instrument(skip_all, fields(log_n = %coset.log_n))]
+    fn selectors_on_coset(&self, coset: Self) -> LagrangeSelectors<Vec<Self::Val>> {
+        let sels = coset
+            .points()
+            .map(|p| self.selectors_at_point(point_to_univariate(p).unwrap()))
+            .collect_vec();
+        LagrangeSelectors {
+            is_first_row: sels.iter().map(|s| s.is_first_row).collect(),
+            is_last_row: sels.iter().map(|s| s.is_last_row).collect(),
+            is_transition: sels.iter().map(|s| s.is_transition).collect(),
+            inv_zeroifier: sels.iter().map(|s| s.inv_zeroifier).collect(),
+        }
+    }
+
+    fn split_domains(&self, num_chunks: usize) -> Vec<Self> {
+        assert!(self.is_standard());
```

Comment:
nit: currently, the only public method to build a `CircleDomain` is through the `standard` method, so this check becomes redundant, if there are no changes in the `CircleDomain`'s API.